sw/host/opentitantool/src/command/spx.rs

```diff
@@ -0,0 +1,171 @@
+use opentitanlib::image::image::Image;
+use opentitanlib::util::file::{FromReader, PemSerilizable, ToWriter};
+
+/// Given the path to a public key, returns the public key. Given
+/// the path to a full keypair, extracts the public key from the private
+/// key and returns the public key.
+fn load_pub_or_priv_key(path: &Path) -> Result<SpxPublicKey> {
+    Ok(match SpxPublicKey::read_pem_file(path) {
+        Ok(pk) => pk,
+        Err(e1) => match SpxKeypair::read_pem_file(path) {
+            Ok(kp) => kp.into_public_key(),
+            Err(e2) => {
+                return Err(anyhow!(
+                    "\n\
+                        Reading as public key: {}\n\
+                        Reading as keypair: {}",
+                    e1,
+                    e2
+                ))
+            }
+        },
+    })
+}
+
+#[derive(Annotate, serde::Serialize)]
+pub struct SpxPublicKeyInfo {
+    pub public_key_num_bits: usize,
+    #[annotate(format=hex)]
+    pub public_key: Vec<u8>,
+}
+
+/// Show public information of a SPHINCS+ public key or key pair.
+#[derive(Debug, StructOpt)]
+pub struct SpxKeyShowCommand {
+    #[structopt(
+        name = "KEY_FILE",
+        help = "SPHINCS+ key file (either just the public key or full keypair)"
+    )]
+    key_file: PathBuf,
+}
+
+impl CommandDispatch for SpxKeyShowCommand {
+    fn run(
+        &self,
+        _context: &dyn Any,
+        _transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        let key = load_pub_or_priv_key(&self.key_file)?;
+
+        Ok(Some(Box::new(SpxPublicKeyInfo {
+            public_key_num_bits: key.pk_len() * 8,
+            public_key: key.pk_as_bytes().to_vec(),
+        })))
+    }
+}
+
+/// Generate a SPHINCS+-SHAKE256-128s-simple public private key pair. The full keypair will be
+/// written to <OUTPUT_DIR>/<BASENAME>.key and the public key will be written to
+/// <OUTPUT_DIR>/<BASENAME>.pub.key.
+#[derive(Debug, StructOpt)]
+pub struct SpxKeyGenerateCommand {
+    #[structopt(name = "OUTPUT_DIR", help = "Output directory")]
+    output_dir: PathBuf,
+    #[structopt(name = "BASENAME", help = "Basename for the generated key pair")]
+    basename: String,
+}
+
+impl CommandDispatch for SpxKeyGenerateCommand {
+    fn run(
+        &self,
+        _context: &dyn Any,
+        _transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        let private_key = SpxKeypair::generate();
+        let mut file = self.output_dir.to_owned();
+        file.push(&self.basename);
+        file.set_extension("pem");
+        private_key.write_pem_file(&file)?;
+
+        file.set_extension("pub.pem");
+        private_key.into_public_key().write_pem_file(&file)?;
+
+        Ok(None)
+    }
+}
+
+#[derive(Debug, StructOpt, CommandDispatch)]
+pub enum SpxKeySubcommands {
+    Show(SpxKeyShowCommand),
+    Generate(SpxKeyGenerateCommand),
+}
+
+#[derive(serde::Serialize)]
+pub struct SpxSignResult {
+    pub signature: String,
```

Comment:
How big is the spx signature?  My recollection is ~8K bytes; do we really want to print it?